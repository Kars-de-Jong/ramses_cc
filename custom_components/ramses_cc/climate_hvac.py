#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
"""Support for Honeywell's RAMSES-II RF protocol, as used by HVAC.

Provides support for climate entities.
"""

import logging
from typing import Any, Dict, Optional

from homeassistant.components.climate import ClimateEntity
from homeassistant.components.climate.const import (
    FAN_AUTO,
    FAN_HIGH,
    FAN_LOW,
    FAN_MEDIUM,
    FAN_OFF,
    ClimateEntityFeature,
    HVACMode,
)
from homeassistant.const import TEMP_CELSIUS  # "Â°C"
from homeassistant.core import callback

from . import EvoZoneBase

_LOGGER = logging.getLogger(__name__)


class RamsesHvac(EvoZoneBase, ClimateEntity):
    """Base for a Honeywell HVAC unit (Fan, HRU, MVHR, PIV, etc)."""

    # PRESET_AWAY = away
    # PRESET_BOOST (timed), 15, 30, 60, other mins
    # PRESET_COMFORT: auto with lower CO2
    # PRESET_NONE: off, low/med/high or auto

    # fan states....
    _attr_fan_modes = [FAN_OFF, FAN_AUTO, FAN_LOW, FAN_MEDIUM, FAN_HIGH]
    _attr_hvac_modes = [HVACMode.HEAT, HVACMode.OFF]
    _attr_supported_features = (
        ClimateEntityFeature.TARGET_TEMPERATURE
        | ClimateEntityFeature.TARGET_HUMIDITY
        | ClimateEntityFeature.FAN_MODE
        | ClimateEntityFeature.PRESET_MODE
    )

    _attr_target_temperature_step = 0.1
    _attr_temperature_unit = TEMP_CELSIUS

    def __init__(self, broker, device) -> None:
        """Initialize a HVAC system."""
        _LOGGER.info("Found a HVAC system: %r", device)

        super().__init__(broker, device)

        self._icon = "mdi:fan"
        self._hvac_modes = None
        self._preset_modes = None
        self._supported_features = None

    @property
    def extra_state_attributes(self) -> Dict[str, Any]:
        """Return the integration-specific state attributes."""
        return {
            **super().extra_state_attributes,
        }

    @property
    def hvac_action(self) -> Optional[str]:
        """Return the Zone's current running hvac operation."""
        return

    @property
    def hvac_mode(self) -> Optional[str]:
        """Return the Zone's hvac operation ie. heat, cool mode."""
        return

    @property
    def preset_mode(self) -> Optional[str]:
        """Return the Zone's current preset mode, e.g., home, away, temp."""
        return

    @callback
    def set_hvac_mode(self, hvac_mode: str) -> None:
        """Set a Zone to one of its native operating modes."""
        return

    @callback
    def set_preset_mode(self, preset_mode: Optional[str]) -> None:
        """Set the preset mode; if None, then revert to following the schedule."""
        return
