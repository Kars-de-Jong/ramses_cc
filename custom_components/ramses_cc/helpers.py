"""Helpers for RAMSES integration."""
from __future__ import annotations

import logging

from homeassistant.const import STATE_UNAVAILABLE, STATE_UNKNOWN
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers import entity_registry as er

from .const import DOMAIN as PLATFORM

OLD_PLATFORM = "evohome_cc" if PLATFORM == "ramses_cc" else "ramses_cc"

_LOGGER = logging.getLogger(__name__)


@callback
def migrate_to_ramses_rf(hass: HomeAssistant, domain: str, unique_id: str):
    """Migrate an entity to the ramses_rf platform (from evohome_rf)."""

    entity_registry = er.async_get(hass)

    if entity_id := entity_registry.async_get_entity_id(
        domain, OLD_PLATFORM, unique_id
    ):
        if (entity := hass.states.get(entity_id)) and entity.state == STATE_UNAVAILABLE:
            hass.states.get(entity_id).state = STATE_UNKNOWN  # HACK

        try:
            entity_registry.async_update_entity_platform(entity_id, PLATFORM)
        except ValueError as exc:
            _LOGGER.error(
                "Migrating %s (%s) to %s failed: %s",
                entity_id,
                unique_id,
                PLATFORM,
                exc,
            )
        else:
            _LOGGER.warning(
                "Migrating %s (%s) to %s: success", entity_id, unique_id, PLATFORM
            )

    # if (
    #     state := hass.states.get(entity_id)
    # ) is not None and state.state != STATE_UNKNOWN:
    #     raise ValueError("Only entities that haven't been loaded can be migrated")
