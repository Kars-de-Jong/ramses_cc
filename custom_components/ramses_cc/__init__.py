"""Support for Honeywell's RAMSES-II RF protocol, as used by CH/DHW & HVAC.

Requires a Honeywell HGI80 (or compatible) gateway.
"""
from __future__ import annotations

from functools import partial
import logging
from typing import Any

from ramses_rf.entity_base import Entity as RamsesRFEntity
from ramses_tx.exceptions import TransportSerialError
import voluptuous as vol

from homeassistant.components.sensor import SensorDeviceClass
from homeassistant.const import (
    ATTR_ID,
    CONF_SCAN_INTERVAL,
    EVENT_HOMEASSISTANT_START,
    Platform,
)
from homeassistant.core import HomeAssistant, ServiceCall, callback
from homeassistant.helpers.dispatcher import async_dispatcher_connect
from homeassistant.helpers.entity import Entity
from homeassistant.helpers.event import async_call_later
from homeassistant.helpers.service import verify_domain_control
from homeassistant.helpers.typing import ConfigType
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator

from .const import BROKER, DOMAIN, SIGNAL_UPDATE
from .coordinator import RamsesBroker
from .schemas import (
    SCH_DOMAIN_CONFIG,
    SVC_SEND_PACKET,
    SVCS_DOMAIN,
    SZ_ADVANCED_FEATURES,
    SZ_MESSAGE_EVENTS,
)

_LOGGER = logging.getLogger(__name__)


CONFIG_SCHEMA = vol.Schema({DOMAIN: SCH_DOMAIN_CONFIG}, extra=vol.ALLOW_EXTRA)

PLATFORMS = [
    Platform.BINARY_SENSOR,
    Platform.CLIMATE,
    Platform.SENSOR,
    Platform.REMOTE,
    Platform.WATER_HEATER,
]


async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
    """Create a ramses_rf (RAMSES_II)-based system."""

    _LOGGER.debug("\r\n\nConfig = %s\r\n", config[DOMAIN])

    broker = RamsesBroker(hass, config)
    try:
        await broker.start()
    except TransportSerialError as exc:
        _LOGGER.error("There is a problem with the serial port: %s", exc)
        return False

    hass.data.setdefault(DOMAIN, {})[BROKER] = broker

    coordinator: DataUpdateCoordinator = DataUpdateCoordinator(
        hass,
        _LOGGER,
        name=DOMAIN,
        update_interval=config[DOMAIN][CONF_SCAN_INTERVAL],
    )
    coordinator.update_method = broker.async_update

    if _LOGGER.isEnabledFor(logging.DEBUG):  # TODO: remove
        app_storage = await broker._async_load_storage()
        _LOGGER.debug("\r\n\nStore = %s\r\n", app_storage)

    # NOTE: .async_listen_once(EVENT_HOMEASSISTANT_START, awaitable_coro)
    # NOTE: will be passed event, as: async def awaitable_coro(_event: Event):
    await coordinator.async_config_entry_first_refresh()  # will save access tokens too
    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, broker.async_update)

    register_domain_services(hass, broker)
    register_domain_events(hass, broker)

    return True


# TODO: add async_ to routines where required to do so
@callback  # TODO: the following is a mess - to add register/deregister of clients
def register_domain_events(hass: HomeAssistant, broker: RamsesBroker) -> None:
    """Set up the handlers for the system-wide events."""

    @callback
    def process_msg(msg, *args, **kwargs):  # process_msg(msg, prev_msg=None)
        if (
            regex := broker.config[SZ_ADVANCED_FEATURES][SZ_MESSAGE_EVENTS]
        ) and regex.match(f"{msg!r}"):
            event_data = {
                "dtm": msg.dtm.isoformat(),
                "src": msg.src.id,
                "dst": msg.dst.id,
                "verb": msg.verb,
                "code": msg.code,
                "payload": msg.payload,
                "packet": str(msg._pkt),
            }
            hass.bus.async_fire(f"{DOMAIN}_message", event_data)

        if broker.learn_device_id and broker.learn_device_id == msg.src.id:
            event_data = {
                "src": msg.src.id,
                "code": msg.code,
                "packet": str(msg._pkt),
            }
            hass.bus.async_fire(f"{DOMAIN}_learn", event_data)

    broker.client.add_msg_handler(process_msg)


@callback  # TODO: add async_ to routines where required to do so
def register_domain_services(hass: HomeAssistant, broker: RamsesBroker):
    """Set up the handlers for the domain-wide services."""

    @verify_domain_control(hass, DOMAIN)
    async def svc_fake_device(call: ServiceCall) -> None:
        try:
            broker.client.fake_device(**call.data)
        except LookupError as exc:
            _LOGGER.error("%s", exc)
            return
        hass.helpers.event.async_call_later(5, broker.async_update)

    @verify_domain_control(hass, DOMAIN)
    async def svc_force_update(_: ServiceCall) -> None:
        await broker.async_update()

    @verify_domain_control(hass, DOMAIN)
    async def svc_send_packet(call: ServiceCall) -> None:
        kwargs = dict(call.data.items())  # is ReadOnlyDict
        if (
            call.data["device_id"] == "18:000730"
            and kwargs.get("from_id", "18:000730") == "18:000730"
            and broker.client.hgi.id
        ):
            kwargs["device_id"] = broker.client.hgi.id
        broker.client.send_cmd(broker.client.create_cmd(**kwargs))
        hass.helpers.event.async_call_later(5, broker.async_update)

    domain_service = SVCS_DOMAIN
    if not broker.config[SZ_ADVANCED_FEATURES].get(SVC_SEND_PACKET):
        del domain_service[SVC_SEND_PACKET]

    services = {k: v for k, v in locals().items() if k.startswith("svc")}
    for name, schema in SVCS_DOMAIN.items():
        if f"svc_{name}" in services:
            hass.services.async_register(
                DOMAIN, name, services[f"svc_{name}"], schema=schema
            )


class RamsesEntity(Entity):
    """Base for any RAMSES II-compatible entity (e.g. Climate, Sensor)."""

    _broker: RamsesBroker
    _device: RamsesRFEntity

    _attr_should_poll = False

    def __init__(self, broker: RamsesBroker, device) -> None:
        """Initialize the entity."""
        self.hass = broker.hass
        self._broker = broker
        self._device = device

        self._attr_unique_id = device.id

    @property
    def extra_state_attributes(self) -> dict[str, Any]:
        """Return the integration-specific state attributes."""
        return {
            ATTR_ID: self._device.id,
        }

    async def async_added_to_hass(self) -> None:
        """Run when entity about to be added to hass."""
        self._broker._entities[self.unique_id] = self
        self.async_on_remove(
            async_dispatcher_connect(
                self.hass, SIGNAL_UPDATE, self.async_write_ha_state
            )
        )

    @callback
    def async_write_ha_state_delayed(self, delay=3) -> None:
        """Write the state to the state machine after a short delay to allow system to quiesce."""
        async_call_later(self.hass, delay, self.async_write_ha_state)


class RamsesSensorBase(RamsesEntity):
    """Base for any Ramses sensor/binary_sensor entity."""

    def __init__(
        self,
        broker,
        device,
        state_attr,
        device_class: SensorDeviceClass | None = None,
        unique_id_attr: str | None = None,
    ) -> None:
        """Initialize the sensor."""
        super().__init__(broker, device)

        self.entity_id = f"{DOMAIN}.{device.id}-{state_attr}"

        self._attr_device_class = device_class
        self._attr_unique_id = f"{device.id}-{unique_id_attr or state_attr}"
        self._state_attr = state_attr

    @property
    def available(self) -> bool:
        """Return True if the sensor is available."""
        return getattr(self._device, self._state_attr) is not None

    @property
    def name(self) -> str:
        """Return the name of the binary_sensor/sensor."""
        if not hasattr(self._device, "name") or not self._device.name:
            return f"{self._device.id} {self._state_attr}"
        return f"{self._device.name} {self._state_attr}"
