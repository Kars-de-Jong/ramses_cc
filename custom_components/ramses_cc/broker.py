"""Broker for RAMSES integration."""
from __future__ import annotations

from datetime import datetime as dt, timedelta
import logging
from threading import Semaphore
from typing import Any

from ramses_rf import Gateway
from ramses_rf.device.base import Device
from ramses_rf.device.hvac import HvacRemoteBase, HvacVentilator
from ramses_rf.entity_base import Entity as RamsesRFEntity
from ramses_rf.schemas import (
    SZ_CONFIG,
    SZ_RESTORE_CACHE,
    SZ_RESTORE_SCHEMA,
    SZ_RESTORE_STATE,
    SZ_SCHEMA,
)
from ramses_rf.system.heat import Evohome, MultiZone, System
from ramses_rf.system.zones import DhwZone, Zone
from ramses_tx.schemas import SZ_PACKET_LOG, SZ_PORT_CONFIG
import voluptuous as vol

from homeassistant.const import CONF_SCAN_INTERVAL, Platform
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.discovery import async_load_platform
from homeassistant.helpers.dispatcher import async_dispatcher_send
from homeassistant.helpers.event import async_call_later, async_track_time_interval
from homeassistant.helpers.storage import Store
from homeassistant.helpers.typing import ConfigType

from .const import DOMAIN, SIGNAL_UPDATE, STORAGE_KEY, STORAGE_VERSION
from .schemas import merge_schemas, normalise_config, schema_is_minimal

_LOGGER = logging.getLogger(__name__)

SZ_CLIENT_STATE = "client_state"
SZ_PACKETS = "packets"
SZ_REMOTES = "remotes"

SAVE_STATE_INTERVAL = timedelta(seconds=300)  # TODO: 5 minutes


class RamsesBroker:
    """Container for client and data."""

    def __init__(self, hass: HomeAssistant, hass_config: ConfigType) -> None:
        """Initialize the client and its data structure(s)."""

        self.hass = hass
        self._store: Store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)

        self.hass_config = hass_config
        _LOGGER.debug("Config = %s", hass_config)
        self._ser_name, self._client_config, self.config = normalise_config(
            hass_config[DOMAIN]
        )

        self.client: Gateway = None  # type: ignore[assignment]
        self._remotes: dict[str, str] = None  # type: ignore[assignment]

        self._services = {}
        self._entities = {}  # domain entities

        # Discovered client objects...
        self._devices: set[Device] = set()
        self._systems: set[System] = set()
        self._zones: set[Zone] = set()

        self._sem = Semaphore(value=1)

        self.learn_device_id = None  # TODO: remove me

    async def start(self) -> None:
        """Invoke the client/co-ordinator (according to the config/cache)."""

        CONFIG_KEYS = (SZ_CONFIG, SZ_PACKET_LOG, SZ_PORT_CONFIG)

        storage = await self._async_load_storage()
        _LOGGER.debug("Storage = %s", storage)

        self._remotes = storage.get(SZ_REMOTES, {}) | self.config[SZ_REMOTES]
        client_state: dict[str, Any] = storage.get(SZ_CLIENT_STATE, {})

        restore_state = self.config[SZ_RESTORE_CACHE][SZ_RESTORE_STATE]
        restore_schema = self.config[SZ_RESTORE_CACHE][SZ_RESTORE_SCHEMA]

        self.client = self._create_client(
            {k: v for k, v in self._client_config.items() if k in CONFIG_KEYS},
            {k: v for k, v in self._client_config.items() if k not in CONFIG_KEYS},
            client_state.get(SZ_SCHEMA, {}) if restore_schema else {},
        )

        cached_packets = (
            self._filter_cached_packets(
                client_state.get(SZ_PACKETS, {}), restore_schema
            )
            if restore_state
            else {}
        )

        await self.client.start(cached_packets=cached_packets)

        # Perform initial update, then poll at intervals
        await self.async_update()
        async_track_time_interval(
            self.hass, self.async_update, self.hass_config[DOMAIN][CONF_SCAN_INTERVAL]
        )

        async_track_time_interval(
            self.hass, self.async_save_client_state, SAVE_STATE_INTERVAL
        )

    def _create_client(
        self,
        client_config: dict[str:Any],
        config_schema: dict[str:Any],
        cached_schema: dict[str:Any] | None = None,
    ) -> Gateway:
        """Create a client with an inital schema (merged or config)."""

        # TODO: move this to volutuous schema
        if not schema_is_minimal(config_schema):  # move this logic into ramses_rf?
            _LOGGER.warning("The config schema is not minimal (consider minimising it)")

        if cached_schema and (merged := merge_schemas(config_schema, cached_schema)):
            try:
                return Gateway(
                    self._ser_name, loop=self.hass.loop, **client_config, **merged
                )
            except (LookupError, vol.MultipleInvalid) as exc:
                # LookupError:     ...in the schema, but also in the block_list
                # MultipleInvalid: ...extra keys not allowed @ data['???']
                _LOGGER.warning("Failed to initialise with merged schema: %s", exc)

        return Gateway(
            self._ser_name, loop=self.hass.loop, **client_config, **config_schema
        )

    def _filter_cached_packets(
        self, cached_packets: dict, restore_schema: bool
    ) -> dict:
        """Filter cached packets for replay on startup."""

        msg_code_filter = ["313F"]
        if not restore_schema:
            msg_code_filter.extend(["0005", "000C"])

        return {
            dtm: pkt
            for dtm, pkt in cached_packets.items()
            if dt.fromisoformat(dtm) > dt.now() - timedelta(days=1)
            and pkt[41:45] not in msg_code_filter
        }

    async def _async_load_storage(self) -> dict:
        """May return an empty dict."""

        app_storage = await self._store.async_load()  # return None if no store
        return app_storage or {}

    async def async_save_client_state(self, *args, **kwargs) -> None:
        """Save the client state to the application store."""

        _LOGGER.info("Saving the client state cache (packets, schema)")

        (schema, packets) = self.client._get_state()
        remotes = self._remotes | {
            k: v._commands for k, v in self._entities.items() if hasattr(v, "_commands")
        }

        await self._store.async_save(
            {
                SZ_CLIENT_STATE: {SZ_SCHEMA: schema, SZ_PACKETS: packets},
                SZ_REMOTES: remotes,
            }
        )

    async def async_update(self, _: dt | None = None) -> None:
        """Retrieve the latest state data from the client library."""

        @callback
        def async_add_devices(platform: str, devices: list[RamsesRFEntity]) -> None:
            if not devices:
                return
            self.hass.async_create_task(
                async_load_platform(
                    self.hass, platform, DOMAIN, {"devices": devices}, self.hass_config
                )
            )

        @callback
        def new_entities(
            known: set[RamsesRFEntity], current: list[RamsesRFEntity]
        ) -> list[RamsesRFEntity]:
            new = current - known
            known |= new
            return new

        new_devices = new_entities(self._devices, set(self.client.devices))
        new_systems = new_entities(self._systems, set(self.client.systems))
        new_zones = new_entities(
            self._zones,
            set().union(
                *[s.zones for s in self.client.systems if isinstance(s, MultiZone)]
            ),
        )

        new_sensor_devices = new_devices | new_systems | new_zones
        async_add_devices(Platform.BINARY_SENSOR, new_sensor_devices)
        async_add_devices(Platform.SENSOR, new_sensor_devices)

        async_add_devices(
            Platform.CLIMATE, [d for d in new_devices if isinstance(d, HvacVentilator)]
        )
        async_add_devices(
            Platform.REMOTE, [d for d in new_devices if isinstance(d, HvacRemoteBase)]
        )
        async_add_devices(
            Platform.CLIMATE, [s for s in new_systems if isinstance(s, Evohome)]
        )

        for zone in new_zones:
            if isinstance(zone, DhwZone):
                async_add_devices(Platform.WATER_HEATER, [zone])
            elif isinstance(zone.tcs, Evohome):
                async_add_devices(Platform.CLIMATE, [zone])

        if new_devices or new_systems or new_zones:
            async_call_later(self.hass, 5, self.async_save_client_state)

        # Trigger state updates of all entities
        async_dispatcher_send(self.hass, SIGNAL_UPDATE)
