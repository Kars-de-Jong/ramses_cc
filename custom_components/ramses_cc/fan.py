#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
"""Support for Honeywell's RAMSES-II RF protocol, as used by CH/DHW & HVAC.

Provides support for climate entities.
"""
import logging
from datetime import datetime as dt
from typing import Any, Dict, Optional

from homeassistant.components.fan import DOMAIN as PLATFORM
from homeassistant.components.fan import FanEntity
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.entity_platform import AddEntitiesCallback, current_platform
from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType

from . import EvoZoneBase
from .const import (
    ATTR_SETPOINT,
    BROKER,
    DATA,
    DOMAIN,
    SERVICE,
    UNIQUE_ID,
    SystemMode,
    ZoneMode,
)
from .schema import (
    CLIMATE_SERVICES,
    CONF_MODE,
    CONF_SYSTEM_MODE,
    SVC_RESET_SYSTEM_MODE,
    SVC_SET_SYSTEM_MODE,
)

_LOGGER = logging.getLogger(__name__)


async def async_setup_platform(
    hass: HomeAssistant,
    config: ConfigType,
    async_add_entities: AddEntitiesCallback,
    discovery_info: DiscoveryInfoType = None,
) -> None:
    """Create the HVAC devices, if any."""
    if discovery_info is None:
        return

    broker = hass.data[DOMAIN][BROKER]

    new_entities = [
        RamsesFan(hass.data[DOMAIN][BROKER], device)
        for device in discovery_info.get("devices", [])
        if hasattr(device, "fan_rate")
    ]

    if new_entities:
        async_add_entities(new_entities)

    if broker.services.get(PLATFORM):
        return
    broker.services[PLATFORM] = True


class RamsesFan(EvoZoneBase, FanEntity):
    """Base for a Honeywell TCS Zone."""

    def __init__(self, broker, device) -> None:
        """Initialize a Zone."""
        _LOGGER.info("Found a Fan: %r", device)
        super().__init__(broker, device)

        self._unique_id = device.id
        self._icon = "mdi:fan"

    @property
    def extra_state_attributes(self) -> Dict[str, Any]:
        """Return the integration-specific state attributes."""
        return {
            "zone_idx": self._device.idx,
            "heating_type": self._device.heating_type,
            "mode": self._device.mode,
            "config": self._device.config,
            **super().extra_state_attributes,
        }
