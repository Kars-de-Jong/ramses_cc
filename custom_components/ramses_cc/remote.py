#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
"""Support for Honeywell's RAMSES-II RF protocol, as used by CH/DHW & HVAC.

Provides support for HVAC RF remotes.
"""

import logging
from typing import Any, Dict, Iterable, Optional

from homeassistant.config_entries import ConfigEntry
from homeassistant.components.remote import DOMAIN as PLATFORM
from homeassistant.components.remote import ATTR_NUM_REPEATS, RemoteEntity
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.typing import DiscoveryInfoType

from . import RamsesDeviceBase as RamsesEntity
from .const import BROKER, DOMAIN

_LOGGER = logging.getLogger(__name__)


async def async_setup_platform(
    hass: HomeAssistant,
    _: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
    discovery_info: DiscoveryInfoType = None,
) -> None:
    """Set up the remote entities.

    discovery_info keys:
      gateway: is the ramses_rf protocol stack (gateway/protocol/transport/serial)
      devices: heat (e.g. CTL, OTB, BDR, TRV) or hvac (e.g. FAN, CO2, REM)
      domains: TCS, DHW and Zones
    """

    if discovery_info is None:
        return

    broker = hass.data[DOMAIN][BROKER]

    new_remotes = [
        RamsesRemote(broker, device)
        for device in discovery_info["remotes"]
    ]  # and (not device._is_faked or device["fakable"])

    async_add_entities(new_remotes)

    if not broker._services.get(PLATFORM) and new_remotes:
        broker._services[PLATFORM] = True


class RamsesRemote(RamsesEntity, RemoteEntity):
    """Representation of a generic sensor."""

    def __init__(self, broker, device, **kwargs) -> None:
        """Initialize a sensor."""
        _LOGGER.info("Found a Remote: %s", device)
        super().__init__(broker, device)

        device_id = device.id

        _attr_name = f"{device.id}_remote"

    @property
    def is_on(self) -> None | bool:
        """Return true if device is on."""
        return not self.coordinator.data.state.standby

    async def async_turn_on(self, **kwargs: Any) -> None:
        """Turn the device on."""
        await self.coordinator.roku.remote("poweron")

    async def async_turn_off(self, **kwargs: Any) -> None:
        """Turn the device off."""
        await self.coordinator.roku.remote("poweroff")

    async def async_send_command(self, command: Iterable[str], **kwargs: Any) -> None:
        """Send a command to one device."""
        num_repeats = kwargs[ATTR_NUM_REPEATS]

        for _ in range(num_repeats):
            for single_command in command:
                await self.coordinator.roku.remote(single_command)
