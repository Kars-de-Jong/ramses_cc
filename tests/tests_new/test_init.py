"""Tests for the ramses_cc tests."""

from __future__ import annotations

from typing import Any

import pytest
from homeassistant.config_entries import ConfigEntryState
from homeassistant.core import HomeAssistant
from homeassistant.setup import async_setup_component
from syrupy import SnapshotAssertion

from custom_components.ramses_cc import DOMAIN

from ..virtual_rf import VirtualRf
from .common import configuration_fixture, storage_fixture
from .const import TEST_SYSTEMS


@pytest.mark.parametrize("instance", TEST_SYSTEMS)
async def test_entities(
    hass: HomeAssistant,
    hass_storage: dict[str, Any],
    instance: str,
    rf: VirtualRf,
    snapshot: SnapshotAssertion,
) -> None:
    """Test State after setup of an instance of the integration."""

    hass_storage[DOMAIN] = storage_fixture(instance)

    config = configuration_fixture(instance)
    config[DOMAIN]["serial_port"] = rf.ports[0]

    assert await async_setup_component(hass, DOMAIN, config)
    await hass.async_block_till_done()

    try:
        entry = hass.config_entries.async_entries(DOMAIN)[0]
        assert entry.state == ConfigEntryState.LOADED

        assert hass.states.async_all(domain_filter=DOMAIN) == snapshot

    finally:  # Prevent non-useful errors in teardown
        assert await hass.config_entries.async_unload(entry.entry_id)
