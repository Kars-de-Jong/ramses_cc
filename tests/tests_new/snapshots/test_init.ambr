# serializer version: 1
# name: test_entities[default]
  list([
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'battery_level': None,
        'device_class': 'battery',
        'friendly_name': 'THM 03:123456 Battery',
        'id': '03:123456',
      }),
      'context': <ANY>,
      'entity_id': 'binary_sensor.03_123456_battery_low',
      'last_changed': <ANY>,
      'last_reported': <ANY>,
      'last_updated': <ANY>,
      'state': 'off',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'problem',
        'friendly_name': 'Controller 01:145038 System status',
      }),
      'context': <ANY>,
      'entity_id': 'binary_sensor.01_145038_status',
      'last_changed': <ANY>,
      'last_reported': <ANY>,
      'last_updated': <ANY>,
      'state': 'unavailable',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'problem',
        'friendly_name': 'Controller 01:145038 Active fault',
      }),
      'context': <ANY>,
      'entity_id': 'binary_sensor.01_145038_active_fault',
      'last_changed': <ANY>,
      'last_reported': <ANY>,
      'last_updated': <ANY>,
      'state': 'unavailable',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'temperature',
        'friendly_name': 'THM 03:123456 Temperature',
        'id': '03:123456',
        'setpoint': None,
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfTemperature.CELSIUS: 'Â°C'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.03_123456_temperature',
      'last_changed': <ANY>,
      'last_reported': <ANY>,
      'last_updated': <ANY>,
      'state': 'unknown',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'friendly_name': 'Controller 01:145038 Heat demand',
        'icon': 'mdi:radiator-off',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': '%',
      }),
      'context': <ANY>,
      'entity_id': 'sensor.01_145038_heat_demand',
      'last_changed': <ANY>,
      'last_reported': <ANY>,
      'last_updated': <ANY>,
      'state': 'unavailable',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'current_temperature': None,
        'friendly_name': 'Controller 01:145038',
        'heat_demand': None,
        'heat_demands': None,
        'hvac_modes': list([
          <HVACMode.HEAT: 'heat'>,
          <HVACMode.OFF: 'off'>,
          <HVACMode.AUTO: 'auto'>,
        ]),
        'icon': 'mdi:thermostat',
        'id': '01:145038',
        'max_temp': None,
        'min_temp': None,
        'preset_mode': None,
        'preset_modes': list([
          'none',
          'away',
          'custom',
          'home',
          'eco',
        ]),
        'relay_demands': None,
        'supported_features': <ClimateEntityFeature: 400>,
        'system_mode': None,
        'tpi_params': None,
      }),
      'context': <ANY>,
      'entity_id': 'climate.01_145038',
      'last_changed': <ANY>,
      'last_reported': <ANY>,
      'last_updated': <ANY>,
      'state': 'unknown',
    }),
  ])
# ---
# name: test_entities[minimal]
  list([
  ])
# ---
# name: test_entities[remotes]
  list([
  ])
# ---
